zerver/lib/mention.py

```diff
@@ -165,15 +180,58 @@ def get_stream_name_map(self, stream_names: set[str]) -> dict[str, int]:
                 .values(
                     "id",
                     "name",
+                    "recipient_id",
+                    "history_public_to_subscribers",
                 )
             )
 
             for row in rows:
-                self.stream_cache[row["name"]] = row["id"]
+                self.stream_cache[row["name"]] = ChannelInfo(
+                    row["id"], row["recipient_id"], row["history_public_to_subscribers"]
+                )
                 result[row["name"]] = row["id"]
 
         return result
 
+    def get_topic_info_map(
+        self, channel_topic: set[ChannelTopicInfo], message_sender: UserProfile | None
+    ) -> dict[ChannelTopicInfo, int]:
+        if not channel_topic:
+            return {}
+
+        result: dict[ChannelTopicInfo, int] = {}
+        unseen_channel_topic: list[ChannelTopicInfo] = []
+
+        for channel_topic_object in channel_topic:
+            if channel_topic_object in self.topic_cache:
+                result[channel_topic_object] = self.topic_cache[channel_topic_object]
+            else:
+                unseen_channel_topic.append(channel_topic_object)
+
+        for channel_topic_object in unseen_channel_topic:
+            channel_info = self.stream_cache.get(channel_topic_object.channel_name)
+
+            assert channel_info is not None
+            recipient_id = channel_info.recipient_id
+            topic_name = channel_topic_object.topic_name
+            history_public_to_subscribers = channel_info.history_public_to_subscribers
+
+            topic_latest_message = get_last_message_for_user_in_topic(
+                self.realm_id,
+                message_sender,
+                recipient_id,
+                topic_name,
+                history_public_to_subscribers,
+            )
+
+            if topic_latest_message is None:
+                continue
```

Comment:
I think it's better to skip the `continue` and store `None` -- caching that we checked a topic and there was nothing is probably better than doing nothing, given this isn't cheap.